### PR TITLE
generic table formatter

### PR DESCRIPTION
**Pull request by poettering**
This adds a generic implementation of a formatter for tables. A bit like libsmartcols, but also totally not. Much simpler (I think) and closer to the code we already have in place (for example, reuses the ellipsation we already have).

I wrote this because I didn't want to add yet another manual, homegrown, awful table formatter for the portable service control tool. I ported loginctl and machinectl over to use the new code, but there's plenty of more code that really should be ported over sooner or later, most prominently systemctl, but also busctl, systemd-inhibit --list, coredumpctl and so on...

### CODE REVIEW

File: src/basic/format-table.c [hunk 1]
```diff
@@ -0,0 +1,1245 @@
...
+
+        assert(n_columns > 0);
+
+        t = new0(Table, 1);
+        if (!t)
+                return NULL;
+
+        t->n_columns = n_columns;
+        t->header = true;
+        t->width = (size_t) -1;
```
Comment (keszybz) [inline src/basic/format-table.c]: I think it'd be nicer to use `new` and a structured initializer.

File: src/basic/format-table.c [hunk 2]
```diff
@@ -0,0 +1,1245 @@
...
+                return NULL;
+
+        d->n_ref = 1;
+        d->type = type;
+        d->minimum_width = minimum_width;
+        d->maximum_width = maximum_width;
+        d->weight = weight;
+        d->align_percent = align_percent;
+        d->ellipsize_percent = ellipsize_percent;
+        memcpy_safe(d->data, data, data_size);
```
Comment (keszybz) [inline src/basic/format-table.c:309]: `malloc` and structured initializer?

Comment (poettering) [reply src/basic/format-table.c:309]: Hmm, this one is tricky... we allocate a object here that might be shorter than the structure's definition (for example if we allocate an uint32_t cell data, which is smaller than uint64_t which is in the union). Hence we shouldn't really initialize more than we actually allocate...

File: src/basic/format-table.c (hunk 1 above)
Comment (poettering) [reply src/basic/format-table.c]: will fix

**Comment by poettering** [comment]
Force pushed a new version, with three changes:

1. use the structure initialize for Table as proposed
2. fixed the memory issue
3. added two TODO list items: one to remind us that DECIMAL_STR_WIDTH() vs. DECIMAL_STR_MAX() is chaos, and we should fix it (this was the problem causing the memory usage), and one to remind us to port other relevant tools over to this formatter too

**Pull request merged**